Conflict when creating a Network ACL connected to Subnet

### Overview
- Conflict when creating a Network ACL connected to Subnet (Subnet 과 연결된 상태에서, Network ACL 생성시 에러)

#### test.tf
```hcl
resource "ncloud_vpc" "vpc" {
	ipv4_cidr_block = "10.3.0.0/16"
}

resource "ncloud_network_acl" "nacl" {
	vpc_no      = ncloud_vpc.vpc.vpc_no
	description = "test acc for network acl"
}

resource "ncloud_subnet" "bar" {
	vpc_no             = ncloud_vpc.vpc.vpc_no
	subnet             = "10.3.0.0/24"
	zone               = "KR-1"
	network_acl_no     = ncloud_network_acl.nacl.id
	subnet_type        = "PUBLIC"
	usage_type         = "GEN"
}

resource "ncloud_network_acl_rule" "nacl_rule_inbound_100" {
	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
	priority          = 100
	protocol          = "TCP"
	rule_action       = "ALLOW"
	port_range        = "22"
	ip_block          = "0.0.0.0/0"
	rule_type         = "INBND"
}

resource "ncloud_network_acl_rule" "nacl_rule_inbound_110" {
	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
	priority          = 110
	protocol          = "TCP"
	rule_action       = "ALLOW"
	port_range        = "80"
	ip_block          = "0.0.0.0/0"
	rule_type         = "INBND"
}
```

#### Log
```log
2020/08/31 21:32:14 [INFO] resource_ncloud_subnet > CreateSubnet params={"regionCode":"KR","subnetTypeCode":"PUBLIC","usageTypeCode":"GEN","networkAclNo":"1116","subnet":"10.3.0.0/24","subnetName":"test-network-acl-rule-vvnuo","vpcNo":"1071","zoneCode":"KR-1"}
2020/08/31 21:32:15 [INFO] resource_ncloud_subnet > CreateSubnet response={"requestId":"8df17fbe-d19d-4e9b-96c1-3a1ae784c70e","returnCode":"0","returnMessage":"success","totalRows":1,"subnetList":[{"subnetNo":"1047","vpcNo":"1071","zoneCode":"KR-1","subnetName":"test-network-acl-rule-vvnuo","subnet":"10.3.0.0/24","subnetStatus":{"code":"CREATING","codeName":"creating"},"createDate":"2020-08-31T21:32:16+0900","subnetType":{"code":"PUBLIC","codeName":"Public"},"usageType":{"code":"GEN","codeName":"General"},"networkAclNo":"1116"}]}
2020/08/31 21:32:15 [INFO] Subnet ID: 1047
...
2020/08/31 21:32:36 [INFO] resource_ncloud_subnet > GetSubnetDetail params={"regionCode":"KR","subnetNo":"1047"}
2020/08/31 21:32:37 [INFO] resource_ncloud_subnet > GetSubnetDetail response={"requestId":"b744d870-2f49-4c16-b400-0bd2542897b2","returnCode":"0","returnMessage":"success","totalRows":1,"subnetList":[{"subnetNo":"1047","vpcNo":"1071","zoneCode":"KR-1","subnetName":"test-network-acl-rule-vvnuo","subnet":"10.3.0.0/24","subnetStatus":{"code":"RUN","codeName":"run"},"createDate":"2020-08-31T21:32:16+0900","subnetType":{"code":"PUBLIC","codeName":"Public"},"usageType":{"code":"GEN","codeName":"General"},"networkAclNo":"1116"}]}
2020/08/31 21:32:37 [INFO] resource_ncloud_network_acl_rule > AddNetworkAclInboundRule params={"regionCode":"KR","networkAclNo":"1116","networkAclRuleList":[{"networkAclRuleDescription":"test-network-acl-rule-vvnuo","ipBlock":"0.0.0.0/0","ruleActionCode":"ALLOW","portRange":"22","priority":100,"protocolTypeCode":"TCP"}]}
2020/08/31 21:32:37 [INFO] resource_ncloud_network_acl_rule > AddNetworkAclInboundRule params={"regionCode":"KR","networkAclNo":"1116","networkAclRuleList":[{"networkAclRuleDescription":"tf-testacc-network-acl-rule","ipBlock":"0.0.0.0/0","ruleActionCode":"ALLOW","portRange":"443","priority":120,"protocolTypeCode":"TCP"}]}
2020/08/31 21:32:37 [INFO] resource_ncloud_network_acl_rule > AddNetworkAclInboundRule params={"regionCode":"KR","networkAclNo":"1116","networkAclRuleList":[{"networkAclRuleDescription":"tf-testacc-network-acl-rule","ipBlock":"0.0.0.0/0","ruleActionCode":"ALLOW","portRange":"80","priority":110,"protocolTypeCode":"TCP"}]}
2020/08/31 21:32:37 [INFO] resource_ncloud_network_acl_rule > AddNetworkAclOutboundRule params={"regionCode":"KR","networkAclNo":"1116","networkAclRuleList":[{"networkAclRuleDescription":"tf-testacc-network-acl-rule","ipBlock":"0.0.0.0/0","ruleActionCode":"ALLOW","portRange":"1-65535","priority":100,"protocolTypeCode":"TCP"}]}
2020/08/31 21:32:37 [ERROR] resource_ncloud_network_acl_rule > AddNetworkAclInboundRule error params={"regionCode":"KR","networkAclNo":"1116","networkAclRuleList":[{"networkAclRuleDescription":"tf-testacc-network-acl-rule","ipBlock":"0.0.0.0/0","ruleActionCode":"ALLOW","portRange":"80","priority":110,"protocolTypeCode":"TCP"}]}, err=Status: 400 Bad Request, Body: {"responseError": {
  "returnCode": "1011002",
  "returnMessage": "You cannot access the appropriate Network ACL."
}}
2020/08/31 21:32:37 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: Status: 400 Bad Request, Body: {"responseError": {
  "returnCode": "1011002",
  "returnMessage": "You cannot access the appropriate Network ACL."
}}
2020/08/31 21:32:37 [ERROR] <root>: eval: *terraform.EvalSequence, err: Status: 400 Bad Request, Body: {"responseError": {
  "returnCode": "1011002",
  "returnMessage": "You cannot access the appropriate Network ACL."
}}
```

### TODO
- [ ] Retrty `AddNetworkAclInboundRule` and `RemoveNetworkAclInboundRule` if got error `returnCode: 1011002`

--- a/ncloud/common.go
+++ b/ncloud/common.go
@@ -22,7 +22,8 @@ const (
 
 	ApiErrorDetachingMountedStorage = "24002"
 
-	ApiErrorAcgCantChangeSameTime = "1007009"
+	ApiErrorAcgCantChangeSameTime           = "1007009"
+	ApiErrorNetworkAclCantAccessaApropriate = "1011002"
 )
 
 const (

--- a/ncloud/resource_ncloud_access_control_group_rule.go
+++ b/ncloud/resource_ncloud_access_control_group_rule.go
@@ -354,7 +354,7 @@ func deleteVpcAccessControlGroupRule(d *schema.ResourceData, config *ProviderCon
 
 	err = resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
 		if rule.RuleType == "INBND" {
-			reqParams := &vserver.RemoveAccessControlGroupInboundRuleRequest{
+			reqParams = &vserver.RemoveAccessControlGroupInboundRuleRequest{
 				RegionCode:                 &config.RegionCode,
 				AccessControlGroupNo:       &rule.AccessControlGroupNo,
 				VpcNo:                      accessControlGroup.VpcNo,
@@ -362,9 +362,9 @@ func deleteVpcAccessControlGroupRule(d *schema.ResourceData, config *ProviderCon
 			}
 
 			logCommonRequest("deleteVpcAccessControlGroupRule", reqParams)
-			resp, err = config.Client.vserver.V2Api.RemoveAccessControlGroupInboundRule(reqParams)
+			resp, err = config.Client.vserver.V2Api.RemoveAccessControlGroupInboundRule(reqParams.(*vserver.RemoveAccessControlGroupInboundRuleRequest))
 		} else {
-			reqParams := &vserver.RemoveAccessControlGroupOutboundRuleRequest{
+			reqParams = &vserver.RemoveAccessControlGroupOutboundRuleRequest{
 				RegionCode:                 &config.RegionCode,
 				AccessControlGroupNo:       &rule.AccessControlGroupNo,
 				VpcNo:                      accessControlGroup.VpcNo,
@@ -372,7 +372,7 @@ func deleteVpcAccessControlGroupRule(d *schema.ResourceData, config *ProviderCon
 			}
 
 			logCommonRequest("deleteVpcAccessControlGroupRule", reqParams)
-			resp, err = config.Client.vserver.V2Api.RemoveAccessControlGroupOutboundRule(reqParams)
+			resp, err = config.Client.vserver.V2Api.RemoveAccessControlGroupOutboundRule(reqParams.(*vserver.RemoveAccessControlGroupOutboundRuleRequest))
 		}
 
 		if err == nil {

--- a/ncloud/resource_ncloud_network_acl_rule.go
+++ b/ncloud/resource_ncloud_network_acl_rule.go
@@ -3,9 +3,11 @@ package ncloud
 import (
 	"bytes"
 	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"log"
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/NaverCloudPlatform/ncloud-sdk-go-v2/ncloud"
 	"github.com/NaverCloudPlatform/ncloud-sdk-go-v2/services/vpc"
@@ -117,41 +119,59 @@ func resourceNcloudNetworkACLRuleCreate(d *schema.ResourceData, meta interface{}
 		networkACLRule.NetworkAclRuleDescription = ncloud.String(v.(string))
 	}
 
-	if d.Get("rule_type").(string) == "INBND" {
-		reqParams := &vpc.AddNetworkAclInboundRuleRequest{
-			RegionCode:         &config.RegionCode,
-			NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
-			NetworkAclRuleList: []*vpc.AddNetworkAclRuleParameter{networkACLRule},
+	var reqParams interface{}
+	var resp interface{}
+
+	err := resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		var err error
+		if d.Get("rule_type").(string) == "INBND" {
+			reqParams = &vpc.AddNetworkAclInboundRuleRequest{
+				RegionCode:         &config.RegionCode,
+				NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
+				NetworkAclRuleList: []*vpc.AddNetworkAclRuleParameter{networkACLRule},
+			}
+
+			logCommonRequest("resource_ncloud_network_acl_rule > AddNetworkAclInboundRule", reqParams)
+			resp, err = config.Client.vpc.V2Api.AddNetworkAclInboundRule(reqParams.(*vpc.AddNetworkAclInboundRuleRequest))
+		} else {
+			reqParams = &vpc.AddNetworkAclOutboundRuleRequest{
+				RegionCode:         &config.RegionCode,
+				NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
+				NetworkAclRuleList: []*vpc.AddNetworkAclRuleParameter{networkACLRule},
+			}
+
+			logCommonRequest("resource_ncloud_network_acl_rule > AddNetworkAclOutboundRule", reqParams)
+			resp, err = config.Client.vpc.V2Api.AddNetworkAclOutboundRule(reqParams.(*vpc.AddNetworkAclOutboundRuleRequest))
 		}
 
-		logCommonRequest("resource_ncloud_network_acl_rule > AddNetworkAclInboundRule", reqParams)
-		resp, err := config.Client.vpc.V2Api.AddNetworkAclInboundRule(reqParams)
-		if err != nil {
-			logErrorResponse("resource_ncloud_network_acl_rule > AddNetworkAclInboundRule", err, reqParams)
-			return err
+		if err == nil {
+			return resource.NonRetryableError(err)
 		}
 
-		logResponse("resource_ncloud_network_acl_rule > AddNetworkAclInboundRule", resp)
-	} else {
-		reqParams := &vpc.AddNetworkAclOutboundRuleRequest{
-			RegionCode:         &config.RegionCode,
-			NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
-			NetworkAclRuleList: []*vpc.AddNetworkAclRuleParameter{networkACLRule},
+		errBody, _ := GetCommonErrorBody(err)
+		if errBody.ReturnCode == ApiErrorNetworkAclCantAccessaApropriate {
+			logErrorResponse("retry AddNetworkAclRule", err, reqParams)
+			time.Sleep(time.Second * 5)
+			return resource.RetryableError(err)
 		}
 
-		logCommonRequest("resource_ncloud_network_acl_rule > AddNetworkAclOutboundRule", reqParams)
-		resp, err := config.Client.vpc.V2Api.AddNetworkAclOutboundRule(reqParams)
-		if err != nil {
-			logErrorResponse("resource_ncloud_network_acl_rule > AddNetworkAclOutboundRule", err, reqParams)
-			return err
-		}
+		return resource.NonRetryableError(err)
+	})
 
-		logResponse("resource_ncloud_network_acl_rule > AddNetworkAclOutboundRule", resp)
+	if err != nil {
+		logErrorResponse("resource_ncloud_network_acl_rule > AddNetworkAclRule", err, reqParams)
+		return err
 	}
+
+	logResponse("resource_ncloud_network_acl_rule > AddNetworkAclRule", resp)
 
 	d.SetId(networkACLIdRuleHash(d.Get("network_acl_no").(string), d.Get("rule_type").(string), d.Get("priority").(int)))
 
 	log.Printf("[INFO] Network ACL Rule ID: %s", d.Id())
+
+	if err = waitForNcloudNetworkACLRunning(config, d.Get("network_acl_no").(string)); err != nil {
+		return err
+	}
 
 	return resourceNcloudNetworkACLRuleRead(d, meta)
 }
@@ -198,36 +218,55 @@ func resourceNcloudNetworkACLRuleDelete(d *schema.ResourceData, meta interface{}
 		ProtocolTypeCode: ncloud.String(d.Get("protocol").(string)),
 	}
 
-	if d.Get("rule_type").(string) == "INBND" {
-		reqParams := &vpc.RemoveNetworkAclInboundRuleRequest{
-			RegionCode:         &config.RegionCode,
-			NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
-			NetworkAclRuleList: []*vpc.RemoveNetworkAclRuleParameter{networkACLRule},
+	var reqParams interface{}
+	var resp interface{}
+
+	err := resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		var err error
+
+		if d.Get("rule_type").(string) == "INBND" {
+			reqParams := &vpc.RemoveNetworkAclInboundRuleRequest{
+				RegionCode:         &config.RegionCode,
+				NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
+				NetworkAclRuleList: []*vpc.RemoveNetworkAclRuleParameter{networkACLRule},
+			}
+
+			logCommonRequest("resource_ncloud_network_acl_rule > RemoveNetworkAclInboundRule", reqParams)
+			resp, err = config.Client.vpc.V2Api.RemoveNetworkAclInboundRule(reqParams)
+		} else {
+			reqParams := &vpc.RemoveNetworkAclOutboundRuleRequest{
+				RegionCode:         &config.RegionCode,
+				NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
+				NetworkAclRuleList: []*vpc.RemoveNetworkAclRuleParameter{networkACLRule},
+			}
+
+			logCommonRequest("resource_ncloud_network_acl_rule > RemoveNetworkAclOutboundRule", reqParams)
+			resp, err = config.Client.vpc.V2Api.RemoveNetworkAclOutboundRule(reqParams)
 		}
 
-		logCommonRequest("resource_ncloud_network_acl_rule > RemoveNetworkAclInboundRule", reqParams)
-		resp, err := config.Client.vpc.V2Api.RemoveNetworkAclInboundRule(reqParams)
-		if err != nil {
-			logErrorResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclInboundRule", err, reqParams)
-			return err
+		if err == nil {
+			return resource.NonRetryableError(err)
 		}
 
-		logResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclInboundRule", resp)
-	} else {
-		reqParams := &vpc.RemoveNetworkAclOutboundRuleRequest{
-			RegionCode:         &config.RegionCode,
-			NetworkAclNo:       ncloud.String(d.Get("network_acl_no").(string)),
-			NetworkAclRuleList: []*vpc.RemoveNetworkAclRuleParameter{networkACLRule},
+		errBody, _ := GetCommonErrorBody(err)
+		if errBody.ReturnCode == ApiErrorNetworkAclCantAccessaApropriate {
+			logErrorResponse("retry RemoveNetworkAclRule", err, reqParams)
+			time.Sleep(time.Second * 5)
+			return resource.RetryableError(err)
 		}
 
-		logCommonRequest("resource_ncloud_network_acl_rule > RemoveNetworkAclOutboundRule", reqParams)
-		resp, err := config.Client.vpc.V2Api.RemoveNetworkAclOutboundRule(reqParams)
-		if err != nil {
-			logErrorResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclOutboundRule", err, reqParams)
-			return err
-		}
+		return resource.NonRetryableError(err)
+	})
 
-		logResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclOutboundRule", resp)
+	if err != nil {
+		logErrorResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclRule", err, reqParams)
+		return err
+	}
+
+	logResponse("resource_ncloud_network_acl_rule > RemoveNetworkAclRule", resp)
+
+	if err = waitForNcloudNetworkACLRunning(config, d.Get("network_acl_no").(string)); err != nil {
+		return err
 	}
 
 	return nil
@@ -258,6 +297,26 @@ func getNetworkACLRuleInstance(d *schema.ResourceData, config *ProviderConfig) (
 	}
 
 	return nil, fmt.Errorf("Not found acl rule, got: %#v", resp.NetworkAclRuleList)
+}
+
+func waitForNcloudNetworkACLRunning(config *ProviderConfig, id string) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"SET"},
+		Target:  []string{"RUN"},
+		Refresh: func() (interface{}, string, error) {
+			instance, err := getNetworkACLInstance(config, id)
+			return VpcCommonStateRefreshFunc(instance, err, "NetworkAclStatus")
+		},
+		Timeout:    DefaultTimeout,
+		Delay:      2 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for Network ACL (%s) to become termintaing: %s", id, err)
+	}
+
+	return nil
 }
 
 func networkACLIdRuleHash(networkACLId string, ruleType string, priority int) string {

--- a/ncloud/resource_ncloud_network_acl_rule_test.go
+++ b/ncloud/resource_ncloud_network_acl_rule_test.go
@@ -48,6 +48,29 @@ func TestAccResourceNcloudNetworkACLRule_basic(t *testing.T) {
 	})
 }
 
+func TestAccResourceNcloudNetworkACLRule_AssociatedSubnet(t *testing.T) {
+	var networkACLRule vpc.NetworkAclRule
+
+	name := fmt.Sprintf("test-nacl-rule-subnet-%s", acctest.RandString(5))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckNetworkACLRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccResourceNcloudNetworkACLRuleConfigAssociatedSubnet(name),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckNetworkACLRuleExists("ncloud_network_acl_rule.nacl_rule_inbound_100", &networkACLRule),
+					testAccCheckNetworkACLRuleExists("ncloud_network_acl_rule.nacl_rule_inbound_110", &networkACLRule),
+					testAccCheckNetworkACLRuleExists("ncloud_network_acl_rule.nacl_rule_inbound_120", &networkACLRule),
+					testAccCheckNetworkACLRuleExists("ncloud_network_acl_rule.nacl_rule_outbound_100", &networkACLRule),
+				),
+			},
+		},
+	})
+}
+
 func TestAccResourceNcloudNetworkACLRule_disappears(t *testing.T) {
 	var networkACLRule vpc.NetworkAclRule
 
@@ -125,6 +148,78 @@ resource "ncloud_network_acl_rule" "nacl_rule_outbound_100" {
 	ip_block          = "0.0.0.0/0"
 	rule_type         = "OTBND"
 	description       = "tf-testacc-network-acl-rule"
+}`, name)
+}
+
+func testAccResourceNcloudNetworkACLRuleConfigAssociatedSubnet(name string) string {
+	return fmt.Sprintf(`
+resource "ncloud_vpc" "vpc" {
+	name            = "%[1]s"
+	ipv4_cidr_block = "10.3.0.0/16"
+}
+
+resource "ncloud_subnet" "subnet" {
+	vpc_no             = ncloud_vpc.vpc.vpc_no
+	name               = "%[1]s"
+	subnet             = "10.3.0.0/24"
+	zone               = "KR-1"
+	network_acl_no     = ncloud_network_acl.nacl.id
+	subnet_type        = "PUBLIC"
+	usage_type         = "GEN"
+}
+
+resource "ncloud_network_acl" "nacl" {
+	vpc_no      = ncloud_vpc.vpc.vpc_no
+	name        = "%[1]s"
+	description = "test acc for network acl"
+}
+
+resource "ncloud_network_acl_rule" "nacl_rule_inbound_100" {
+	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
+	priority          = 100
+	protocol          = "TCP"
+	rule_action       = "ALLOW"
+	port_range        = "22"
+	ip_block          = "0.0.0.0/0"
+	rule_type         = "INBND"
+	description       = "%[1]s"
+	depends_on        = [ncloud_subnet.subnet]
+}
+
+resource "ncloud_network_acl_rule" "nacl_rule_inbound_110" {
+	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
+	priority          = 110
+	protocol          = "TCP"
+	rule_action       = "ALLOW"
+	port_range        = "80"
+	ip_block          = "0.0.0.0/0"
+	rule_type         = "INBND"
+	description       = "tf-testacc-network-acl-rule"
+	depends_on        = [ncloud_subnet.subnet]
+}
+
+resource "ncloud_network_acl_rule" "nacl_rule_inbound_120" {
+	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
+	priority          = 120
+	protocol          = "TCP"
+	rule_action       = "ALLOW"
+	port_range        = "443"
+	ip_block          = "0.0.0.0/0"
+	rule_type         = "INBND"
+	description       = "tf-testacc-network-acl-rule"
+	depends_on        = [ncloud_subnet.subnet]
+}
+
+resource "ncloud_network_acl_rule" "nacl_rule_outbound_100" {
+	network_acl_no    = ncloud_network_acl.nacl.network_acl_no
+	priority          = 100
+	protocol          = "TCP"
+	rule_action       = "ALLOW"
+	port_range        = "1-65535"
+	ip_block          = "0.0.0.0/0"
+	rule_type         = "OTBND"
+	description       = "tf-testacc-network-acl-rule"
+	depends_on        = [ncloud_subnet.subnet]
 }`, name)
 }
 

